Mathlib/Topology/Algebra/Group/Profinite/Defs.lean

```diff
@@ -0,0 +1,149 @@
+/-
+Copyright (c) 2024 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Nailin Guan, Yuyang Zhao, Youle Fang, Yongle Hu
+-/
+import Mathlib.Algebra.Category.Grp.Basic
+import Mathlib.Topology.Category.Profinite.Basic
+import Mathlib.Topology.Algebra.ContinuousMonoidHom
+
+/-!
+
+# Category of Profinite Groups
+
+We say `G` is a profinite group if it is a topological group which is compact and totally
+disconnected.
+
+## Main definitions and results
+
+* `ProfiniteGrp` is the type of profinite groups.
+
+* `FiniteGrp` is the type of finite groups.
+
+* `Pi.profiniteGrp` : The product of profinite groups is also a profinite group.
+
+-/
+
+suppress_compilation
+
+universe u v
+
+open CategoryTheory Topology
+
+/--Defining the structure of profinite group-/
+@[pp_with_univ]
+structure ProfiniteGrp where
+  /--the underlying set is profinite-/
+  toProfinite : Profinite
+  /--it is also a topological group with the topology given-/
+  [isGroup : Group toProfinite]
+  [isTopologicalGroup : TopologicalGroup toProfinite]
+
+/--Defining the structure of finite group-/
+@[pp_with_univ]
+structure FiniteGrp where
+  /--A group that is finite-/
+  toGrp : Grp
+  [isFinite : Finite toGrp]
+
+namespace FiniteGrp
+
+instance : CoeSort FiniteGrp.{u} (Type u) where
+  coe G := G.toGrp
+
+instance : Category FiniteGrp := InducedCategory.category FiniteGrp.toGrp
+
+instance : ConcreteCategory FiniteGrp := InducedCategory.concreteCategory FiniteGrp.toGrp
+
+instance (G : FiniteGrp) : Group G := inferInstanceAs <| Group G.toGrp
+
+instance (G : FiniteGrp) : Finite G := G.isFinite
+
+instance (G H : FiniteGrp) : FunLike (G ⟶ H) G H :=
+  inferInstanceAs <| FunLike (G →* H) G H
+
+instance (G H : FiniteGrp) : MonoidHomClass (G ⟶ H) G H :=
+  inferInstanceAs <| MonoidHomClass (G →* H) G H
+
+/--Making a finite group into a FiniteGrp-/
+def of (G : Type u) [Group G] [Finite G] : FiniteGrp where
+  toGrp := Grp.of G
+  isFinite := ‹_›
+
+/--The morphisms between FiniteGrp-/
+def ofHom {X Y : Type u} [Group X] [Finite X] [Group Y] [Finite Y] (f : X →* Y) : of X ⟶ of Y :=
+  Grp.ofHom f
+
+lemma ofHom_apply {X Y : Type u} [Group X] [Finite X] [Group Y] [Finite Y] (f : X →* Y) (x : X) :
+    ofHom f x = f x :=
+  rfl
+
+end FiniteGrp
+
+namespace ProfiniteGrp
+
+instance : CoeSort ProfiniteGrp (Type u) where
+  coe G := G.toProfinite
```

Comment:
Maybe you should add `coe_of`, `coe_id`, `coe_comp` simp lemmas. See the file `Topology.Category.CompHausLike.Basic` for a template.